hack/lib/golang.sh

```diff
@@ -629,9 +629,9 @@ kube::golang::setup_gomaxprocs() {
       GOTOOLCHAIN="$(kube::golang::hack_tools_gotoolchain)" go -C "${KUBE_ROOT}/hack/tools" install ./ncpu || echo "Will not automatically set GOMAXPROCS"
     fi
     if command -v ncpu >/dev/null 2>&1; then
-      GOMAXPROCS=$(ncpu)
+      GOMAXPROCS=$(( $(ncpu) * ${KUBE_GOMAXPROCS_SCALE_FACTOR:-1} ))
```

Comment:
is not better to use the -parallel flag instead of relying on modifying the environment variable?

Reply:
Depends on whether we want some generic mechanism that could also work elsewhere without `go test`.

FWIW, `KUBE_INTEGRATION_TEST_MAX_CONCURRENCY` also was implemented by setting `GOMAXPROCS`, not with `-parallel`.

Reply:
I think we should look at using parallel when exceeding the cores available. GOMAXPROCS influences the entire runtime and affects how time is set aside to eg service waiters IIRC, setting GOMAXPROCS high lies to the scheduler about available CPU, but we can tell go to run more tests per core which will still schedule to those per-core workers.

Reply:
IE we want to increase the test parallelism but not give go a bad estimate about how to schedule runtime internals.

Reply:
That's a valid point. Let me see whether I can make it `go test` specific via `-parallel`.

Reply:
BTW, does `go test` then run each test with a reduced GOMAXPROCS? When I have 36 tests running in parallel and each assumes that it can use 36 cores, it would massively oversubscribe the machine. Hmm, I just tried. The individual test process has a GOMAXPROCS env variable in its `/proc/<pid>/environment` - but it's 36!

Reply:
I lost my draft comment with more details :(

GOMAXPROCS env should be inherited through the process tree, and go will only default it to the number of ~physical cores if that's not set.

We should probably not set parallel test runs terribly high.

Worth noting: While not all of the details of GOMAXPROCS are well documented, it's clearly stated that this influences the *user threads* running, and not blocked system I/O (which is allowed to be blocked in the background exceeding GOMAXPROCS)